Handle missing files & placeholder docs in the PackageTableView

robyngit opened the pull request:
This PR updates MetacatUI to better handle upcoming changes to the indexer. Next week, the indexer will start creating placeholder documents for files that are referenced in resource maps but are not yet indexed or are missing from the repository. These changes make sure that the `DataPackageView` reflect these placeholder files correctly, and improve error handling and UI feedback for non-existent files.

When the DataPackage collection fetches members from Solr it now includes the `isPlaceHolder_b` field in the query. When one or more documents are marked as placeholders, then the collection attempts to fetch their system metadata to get information for the UI. As long as the file exists, the table looks and behaves normally.

If the file does not exist (404), then the UI indicates that the file is missing and inactivates the download button with a tooltip explaining why (in view mode) or shows an error with directions to replace or remove the file (in edit mode).

Additionally, the "download all" button is inactivated since BagIt returns an error for packages with missing files.

### Other minor improvements:

- in `DataPackage`, avoid retrying member model fetches on certain HTTP status codes (e.g. 401, 404) and record error status and messages on models immediately.
- in `DataONEObject`, add `fileDoesNotExist()` method to check for 404 status.
- Standardized inactive behaviour of `DownloadButtonView` in a single method call (useful for external views like `DataItemView`).
- Fixed duplicate listener and rendering issues in `DataPackageView` and `DataItemView`

### For testing:

This package contains a file that is a placeholder doc, where the file doesn't exist on metacat: https://dev.nceas.ucsb.edu/view/urn%3Auuid%3A243bc6ce-92dc-42d9-948e-c2e78da4b885 (ask Jing for permission to edit to test it in edit mode)

robyngit committed:
Fetch sysmeta if file is placeholder doc in solr

Add isPlaceHolder_b to getDataPackage query
When the index has minimal, placeholder info for a file after fetching a data package from Solr, fetch the full system metadata for that file instead.

Issue #2760

robyngit committed:
Standardize inactive behaviour of download button

Make inactivate method to consistently disable the button, remove href,
set cursor to default, and add tooltip.

Issue #2760

robyngit committed:
Don't refetch pkg members if err is 401, 404, etc

Set the error status code & message on the model immediately when a fetch fails

Issue #2760

robyngit committed:
Handle missing files in DataItemView

Set the same re-render listeners at the start of the render function so it applies to both the "edit" and "view" modes. Allows view to update when file is found or error message is set.

Issue #2760

robyngit committed:
Minor tidying of Package Table related code

- better document "archived" option in QueryService
- trim sysMeta response before checking first character in DataONEObject.parse
- remove redundant code in DataPackageView. Prevent double setting a listener.

Issue #2760

robyngit committed:
Prevent download all when files missing from pkg

Issue #2760

robyngit committed:
Fix variable name error in retry logic of dataPkg

Issue #2760

rushirajnenuji reviewed (approved):
Hey @robyngit  - thanks, this looks great! 🎉

One very minor thing I noticed that may be we can add in the upcoming releases -- the package table shows a nice tooltip with the downloads button disabled (`This file does not exist...`). May be we can add the same for the entities table? Right now, it shows `Error` if the download fails. 

Thank you again @robyngit! 🚀

robyngit committed:
Deactivate entity tbl download btn if file missing

Also log a warning if there's an error in the renderMetadata try-catch block because it may capture unexpected errors (not just network errors)

Issue #2760

robyngit commented:
Yes! Good catch @rushirajnenuji, thank you!! Added a commit that deactivates the download button in the entity table if a file doesn't exist.
